content/blog/2022-05-20-Banked_Memory_Compiler_Backend_in_MLIR/index.md

```diff
@@ -0,0 +1,302 @@
+```mlir
+amc.memory @test1(!amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>, !amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>) {
+  // Specify memory allocation.
+  %0 = amc.alloc : !amc.memref<1024xf32, bank [2]>
+
+  // Specify ports.
+  %1 = amc.create_port(%0 : !amc.memref<1024xf32, bank [2]>) banks [0] : !amc.port<512xf32, r, 1>
+  %2 = amc.create_port(%0 : !amc.memref<1024xf32, bank [2]>) banks [0] : !amc.port<512xf32, w, 1>
+  %3 = amc.create_port(%0 : !amc.memref<1024xf32, bank [2]>) banks [1] : !amc.port<512xf32, r, 1>
+  %4 = amc.create_port(%0 : !amc.memref<1024xf32, bank [2]>) banks [1] : !amc.port<512xf32, w, 1>
+ 
+  // Specify I/O mapping.
+  amc.extern %1, %2, %3, %4 : !amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>, !amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>
+}
+```
+### Lowering
+
+The general lowering pipeline can be view as follows:
+
+Handshake Lowering -> Merge Lowering -> Bank Lowering -> Optimization -> AmcToCalyx
+
+Our general goal here is to perform the straight forward lowering passes first, and then optimize later one we have fully expanded the memory structure. We will dive into each of these lowering passes below.
+
+### Bank Lowering
+
+We will start with the simplest lowering step. This should be viewed as the final lowering pass pre-optimization and simply translated a banked memref into multiple unbanked memref. This step requires that all create ports only access a single bank and create only latency-sensitive ports.
+
+Here is an example of bank lowering:
+```mlir
+// Input
+amc.memory @test1(!amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>, !amc.port<512xf32, r, 1>, !amc.port<512xf32,
+  %0 = amc.alloc : !amc.memref<1024xf32, bank [2]>
+  %1 = amc.create_port(%0 : !amc.memref<1024xf32, bank [2]>) banks [0] : !amc.port<512xf32, r, 1>
+  %2 = amc.create_port(%0 : !amc.memref<1024xf32, bank [2]>) banks [0] : !amc.port<512xf32, w, 1>
+  %3 = amc.create_port(%0 : !amc.memref<1024xf32, bank [2]>) banks [1] : !amc.port<512xf32, r, 1>
+  %4 = amc.create_port(%0 : !amc.memref<1024xf32, bank [2]>) banks [1] : !amc.port<512xf32, w, 1>
+   
+  amc.extern %1, %2, %3, %4 : !amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>, !amc.port<512xf32, r, 1>, !amc.p
+}
+
+// Output after bank lowering
+amc.memory @test1(!amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>, !amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>) {
+  %0 = amc.alloc : !amc.memref<512xf32>
+  %1 = amc.alloc : !amc.memref<512xf32>
+  %2 = amc.create_port(%0 : !amc.memref<512xf32>) : !amc.port<512xf32, r, 1>
+  %3 = amc.create_port(%0 : !amc.memref<512xf32>) : !amc.port<512xf32, w, 1>
+  %4 = amc.create_port(%1 : !amc.memref<512xf32>) : !amc.port<512xf32, r, 1>
+  %5 = amc.create_port(%1 : !amc.memref<512xf32>) : !amc.port<512xf32, w, 1>
+  amc.extern %2, %3, %4, %5 : !amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>, !amc.port<512xf32, r, 1>, !amc.port<512xf32, w, 1>
+}
+```
```

Comment:
Just a sentence or so of context after this example could help readers understand what they are supposed to pay attention to. Namely, `%0` used to be a single 2-banked memory; now it is split into `%0` and `%1`, which are both "plain" (unbanked) memories. The ports have been reassigned to those simpler memories. (Adding explicit descriptions like this can make examples worth 1000x more than they would be alone.)